Dev4Agriculture.ISO11783.ISOXML/Analysis/ISODeviceAnalysis.cs

```diff
@@ -204,6 +229,104 @@ public List<TaskDDIEntry> FindDeviceElementsForDDI(ISOTask isoTask, ushort ddi)
+                };
+                DateTime? startTimestamp = null;
+                foreach (var log in logs)
+                {
+                    foreach (var logData in log.Data)
+                    {
+                        if (logData.HasValue)
+                        {
+                            if (startTimestamp == null && logData.DDIValue == 1)
+                            {
+                                info.StartDate ??= logData.TimeStamp;
+                                startTimestamp = logData.TimeStamp;
+                            }
+                            else if(startTimestamp != null && logData.DDIValue == 0)
+                            {
+                                info.Duration += (logData.TimeStamp - startTimestamp.Value).TotalSeconds;
+                                startTimestamp = null;
+                                info.EndDate = logData.TimeStamp;
+                            }
+                        }
+                    }
+                }
+                elementWorkTime.Add(info);
+            }
+
+            var workingElement = elementWorkTime.OrderByDescending(s => s.Duration).FirstOrDefault();
+            result.DeviceElementId = workingElement.DdiEntry.DeviceElementId;
+            result.DurationInSeconds = workingElement.Duration;
+            result.StartDateTime = workingElement.StartDate.GetValueOrDefault();
+            result.StopDateTime = workingElement.EndDate.GetValueOrDefault();
+
+            var device = _isoxml.Data.Device.FirstOrDefault(s => s.DeviceElement.Any(s => s.DeviceElementId == workingElement.DdiEntry.DeviceElementId));
+            result.DeviceId = device.DeviceId;
+            if (workingElement.DdiEntry.Type == DDIValueType.Property)
+            {
+                var properties = device.DeviceProperty.Where(s => s.DevicePropertyDDI == Utils.FormatDDI(179));
+                var elementDevice = device.DeviceElement.FirstOrDefault(s => s.DeviceElementId == workingElement.DdiEntry.DeviceElementId);
+                var property = properties.FirstOrDefault(s => elementDevice.DeviceObjectReference.Any(dor => dor.DeviceObjectId == s.DevicePropertyObjectId));
+                result.CulturalPractice = (CulturalPracticesType)property.DevicePropertyValue;
+                result.IsPropertySource = true;
+            }
+            else if (isoTask.TryGetLastValue((ushort)DDIList.ActualCulturalPractice, workingElement.DdiEntry.DeviceElementNo(), out var ddiValue, false))
+            {
+                result.CulturalPractice = (CulturalPracticesType)ddiValue;
+            }
+            else
+            {
+                var client = new ClientName(device.ClientNAME);
+                result.CulturalPractice = MapDeviceClassToPracticeType(client.DeviceClass);
+            }
+            return result;
+        }
+
+        private CulturalPracticesType MapDeviceClassToPracticeType(DeviceClass className) => className switch
```

Comment:
This can be usefull as a public function, but I would move it to a Utils Class. What do you think?